rfd/0107-desktop-access-performance-research.md

```diff
@@ -0,0 +1,95 @@
+
+The way it works is both the server and the client has a bitmap cache, but it's the server responsibility to control the contents of both bitmap caches. The server can instruct the client to store bitmap data inside a bitmap cache. It can also tell the client to load stored data from the cache and render in on the screen. The main benefit of this architecture is that the stored bitmap data in the bitmap cache can be loaded and rendered multiple times without the need to send data over again. Only the render bitmap operation from the cache needs have to be transmited over the wire. On top of that single message from the RDP server can contain tens or even hundreds of drawing operations. It can significantly reduce the latency and bandwidth usage.
+
+There are two types of caches specified in the extension. The first one, called `Revision 1` only support memory-based caching and defaults to 6MB of cache storage for 32bpp bitmaps. The other one, called `Revision 2` support persistent disk caching in addition to memory caching and defaults to 40MB of cache storage for 32bpp bitmaps.
+Our choice for a support of caching is to support `Revision 2` cache since it allows to store more data inside its cache.
+
+#### How it works on the protocol level?
+
+Two new types of messages needs to be handled by our library in order to enable support for bitmap caching. Message which stores bitmap data inside a bitmap cache, called `Cache Bitmap - Revision 2` and message which renders stored bitmap data in the bitmap cache, called `MemBlt`.
+Message which stores bitmap data inside a bitmap cache contains bitmap data and cache index generated by the server. On the other hand, operation that renders the bitmap only has cache index and position on the screen where to render the bitmap.
+The flow of the messages is controlled by the server so it is not possible to have a message which renders a bitmap that uses cache key that hasn't been yet used to store a bitmap. 
+
+#### What changes are required to support caching?
+
+Teleport Desktop Access protocol, TDP, needs to be extended with two messages: 
+
+#### 29 - bitmap cache store
+
+This message is sent from the server to the client to store a bitmap in the bitmap cache
+
+```
+| message type (29) | cache_id uint32 | cache_index uint32 | data_length uint32 | data []byte |
+```
+
+#### 30 - bitmap cache load
+
+This message is sent from the server to the client to render a bitmap stored in the bitmap cache
+
+```
+| message type (30) | cache_id uint32 | cache_index uint32 | top uint32 | left uint32 | right uint32 | bottom uint32 |
+```
+
+Update to the UI is also required. The browser memory and javascript's array of hash tables, where cache id is the index of the array, and the cache index is a key in the hash table, will be used to store bitmap data.
```

Comment:
Did you consider alternatives? Is a 40MB array a better option than something like IndexedDB?

Reply:
I don't understand how the 40MB calculation is being made for the V2 cache. 

Looking at [2.2.7.1.4.2 Revision 2 (TS_BITMAPCACHE_CAPABILITYSET_REV2)](https://learn.microsoft.com/pt-br/openspecs/windows_protocols/ms-rdpbcgr/a5b9b9a6-5f67-4089-a95d-009bc8e25bfc), I see the ability to add up to 5 caches (`BitmapCacheNCellInfo`s), each with different limits for the number of entries. Adding up all of the maximum limits I get

600 + 600 + 65536 + 4096 + 2048 = 72880 entries

Assuming each entry means an (up to) 64x64 pixel chunk of the screen, at 32bpp that gives us 

(72880 entries) * (64x64 pixels/entry) * 32 (bits/pixel) = 9552527360 bits

which is 1.19 GB.

Reply:
Yes, I know, but at the same time, in the same specification they say something different here: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/2bf92588-42bd-4527-8b3e-b90c56e292d2

> The Revision 1 bitmap cache defaults to 1500 KB of memory cache for 8 bpp, 3000 KB for 16 bpp, 4500 KB for 24 bpp, and 6000 KB for 32 bpp. The Revision 2 bitmap cache defaults to 10 MB of persistent cache storage for 8 bpp, 20 MB for 16 bpp, 30 MB for 24 bpp, and 40 MB for 32 bpp. The actual layout of the bitmap caches is specified in the Bitmap Cache Capability Sets ([MS-RDPBCGR] section [2.2.7.1.4](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/49ab3cb4-d51e-4fb4-a5a4-1beccc3cce6d)).

Reply:
Just to add to the above, we can control the number of cache entries that we support by changing cache numbers in the capability that you've mentioned and the server will respect that numbers.

Reply:
Ah, I see. I suppose it might take some experimentation to figure out which it actually adheres to. Probably not worth the effort considering that we're not moving forward with the caching.